adds guidelines for error logging structure (#31)
giantswarm/fmt: go/log_messages.md

# Log Messages

## General Guidelines

- We use `%#q` for string formatting in our log & error messages. This prevents
  from escaping when they are embedded in JSON string value.

## Resource Reconciliation

When doing resource reconciliation action we often have to find out the current
state, and apply changes to bring it towards desired state. This often means
calling some API and then emitting debug logs about that. We should try to align

> this logs to become more comfortable when reading them and find problems faster.

Reviewer: ```suggestion
these logs to become more comfortable when reading them so we can find problems faster.
```

Also we should aim for them to be understandable not only by developers but also
by SREs, SEs and any other people who look at them and not necessarily
contribute to the codebase.

### Control Flow

We use `*context` packages for control flow of the reconciliation loop. Here are
log messages associated with specific actions.

```go
r.logger.LogCtx(ctx, "level", "debug", "message", "keeping finalizers")
finalizerskeptcontext.SetKept(ctx)

r.logger.LogCtx(ctx, "level", "debug", "message", "canceling reconciliation")
reconciliationcanceledcontext.SetCanceled(ctx)

r.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource")
resourcecanceledcontext.SetCanceled(ctx)
```

### Resource Intercommunication

> We use `controllercontext` package to pass information between resources.

Reviewer: ```suggestion
We use the `controllercontext` package to pass information between resources.
```

> Because of asynchronous nature of resources the information is not available

Reviewer: ```suggestion
Because of the asynchronous nature of resources the information is not available
```

immediately. If not, the resource should log that and cancel itself.

```go
controllerCtx, err := controllercontext.FromContext(ctx)
if err != nil {
	return microerror.Mask(err)
}

workerASGName := controllerCtx.Status.Drainer.WorkerASGName
if workerASGName == "" {
	r.logger.LogCtx(ctx, "level", "debug", "message", "worker ASG name is not available yet")
	r.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource")
	return nil
}
```

### Error Handling

At the root of error handling code when there is no error to be returned to the
original caller, errors have to be logged. We try to format such logs in the way

> as described below in the example.

Reviewer: ```suggestion
as described in the example below.
```

```go

> s.logger.Log("level", "error", "message", "failed collecting metrics", "stack", fmt.Sprintf("%#v", microerror.Mask(err)))

Reviewer 2: ```suggestion
s.logger.LogCtx(ctx, "level", "error", "message", "failed to collect metrics", "stack", fmt.Sprintf("%#v", microerror.Mask(err)))
```

I think using `LogCtx` in the examples is a good idea, because they are copy/pasted. I suggested `failed to collect metrics`. This is the way I logged errors so far. This is my preference but previous form is also ok with me.

```

The fact gathering statement expresses what happens and in case the action fails
`failed` is prepended to it in the error log message. The additional `stack`
field is added to get some tracing information from the carried error.

```go
s.logger.Log("level", "debug", "message", "collecting metrics")

for _, c := range s.collectors {
	err := c.Collect(ch)
	if err != nil {

> 		s.logger.Log("level", "error", "message", "failed collecting metrics", "stack", fmt.Sprintf("%#v", microerror.Mask(err)))

Reviewer 2: ```suggestion
		s.logger.LogCtx(ctx, "level", "error", "message", "failed to collect metrics", "stack", fmt.Sprintf("%#v", microerror.Mask(err)))
```

	}
}

> s.logger.Log("level", "debug", "message", "collected metrics")

Reviewer 2: ```suggestion
s.logger.LogCtx(ctx, "level", "debug", "message", "collected metrics")
```
